airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/fileformatparser.py

```diff
@@ -132,49 +139,116 @@ def is_binary(self):
+        }
 
     def _convert_options(self, json_schema: Mapping[str, Any] = None):
         """
         https://arrow.apache.org/docs/python/generated/pyarrow.csv.ConvertOptions.html
 
+        build ConvertOptions object like: pa.csv.ConvertOptions(**self._convert_options())
+
         :param json_schema: if this is passed in, pyarrow will attempt to enforce this schema on read, defaults to None
         """
         check_utf8 = True if self._format.get("encoding", "utf8").lower().replace("-", "") == "utf8" else False
         convert_schema = self.json_schema_to_pyarrow_schema(json_schema) if json_schema is not None else None
-        return pa.csv.ConvertOptions(
-            check_utf8=check_utf8, column_types=convert_schema, **json.loads(self._format.get("additional_reader_options", "{}"))
-        )
+        return {
+            **{"check_utf8": check_utf8, "column_types": convert_schema},
+            **json.loads(self._format.get("additional_reader_options", "{}")),
+        }
 
     def get_inferred_schema(self, file: Union[TextIO, BinaryIO]) -> dict:
         """
         https://arrow.apache.org/docs/python/generated/pyarrow.csv.open_csv.html
-        Note: this reads just the first block (as defined in _read_options() block_size) to infer the schema
+
+        This now uses multiprocessing in order to timeout the schema inference as it can hang.
+        Since the hanging code is resistant to signal interrupts, threading/futures doesn't help so needed to multiprocess.
+        https://issues.apache.org/jira/browse/ARROW-11853?page=com.atlassian.jira.plugin.system.issuetabpanels%3Aall-tabpanel
         """
-        streaming_reader = pa_csv.open_csv(file, self._read_options(), self._parse_options(), self._convert_options())
-        schema_dict = {field.name: field.type for field in streaming_reader.schema}
+
+        def get_schema(file_sample: str, read_opts: dict, parse_opts: dict, convert_opts: dict) -> dict:
+            """
+            we need to reimport here to be functional on Windows systems since it doesn't have fork()
+            https://docs.python.org/3.7/library/multiprocessing.html#contexts-and-start-methods
+            """
+            import tempfile
+
+            import pyarrow as pa
+
+            # writing our file_sample to a temporary file to then read in and schema infer as before
+            with tempfile.TemporaryFile() as fp:
+                fp.write(file_sample)
+                fp.seek(0)
+                streaming_reader = pa.csv.open_csv(
+                    fp, pa.csv.ReadOptions(**read_opts), pa.csv.ParseOptions(**parse_opts), pa.csv.ConvertOptions(**convert_opts)
+                )
+                schema_dict = {field.name: field.type for field in streaming_reader.schema}
+            return schema_dict
+
+        # boto3 stuff can't be pickled and so we can't multiprocess with the actual fileobject on Windows systems
```

Comment:
what's a better word than "stuff" ? :P

Reply:
you know... stuff... like things and whatever... 😅
Made this more clear!